specification.md

```diff
@@ -2474,6 +2530,134 @@ authenticity of the downloaded file, for example can be the
+of the messages. If the message type is not recognized, the message can be ignored.
+
+A CustomMessage identifies the type of message and contains binary data with the contents
+of the message. The format of the data will depend on the message type and is outside of
+the scope of the OpAMP protocol.
+
+```protobuf
+message CustomMessage {
+    string type = 1;
+    bytes data = 2;
+}
+```
+
+##### CustomMessage.type
+
+The message type should a reverse FQDN that uniquely identifies the custom message type.
+
+##### CustomMessage.data
+
+Binary data of the message. The Agent and Server must agree on the format of the contents.
+
+#### Examples
+
+The following examples describe possible uses of CustomMessage but are not intended to be
+part of any specification. They only describe how CustomMessage might be used. In these
+examples, "io.opentelemetry." is used to create the FQDN of the message types.
+
+##### Pause/Resume
+
+Suppose an Agent supports the ability to pause and resume collection. When paused, no
+telemetry data is collected or sent. Resume will resume the collection and sending of
+telemetry data. To allow a Server to control this behavior, a CustomMessage could be used.
+
+###### Agent Connection
+
+On connection, the Agent sends a CustomMessageCapabilities message including the message
+types "io.opentelemetry.pause" and "io.opentelemetry.resume". In response, the Server
+sends a CustomMessageCapabilities message including "io.opentelemetry.paused" and
+"io.opentelemetry.resumed".
+
+###### Pause
+
+Server sends a ServerToAgent message containing a CustomMessage with type
+"io.opentelemetry.pause". No data is sent because this is a simple command with no
+additional information.
+
+If the Agent supports this message and is able to successfully pause, it returns an
+AgentToServer message containing a CustomMessage with type "io.opentelemetry.paused" with
+empty data.
+
+If the Agent supports this message but encounters an error trying to pause, it returns an
+AgentToServer message containing a CustomMessage with type "io.opentelemetry.paused" and a
+binary JSON-encoded response containing an error message.
+
+###### Resume
+
+Similar to Pause but with a CustomMessage types "io.opentelemetry.resume" and
+"io.opentelemetry.resumed".
+
+##### Service Discovery
+
+Service discovery involves discovering running services that are accessible to the Agent
+for which the Agent could collect telemetry. In this example the Server will send a
+message to the Agent to ask for available services. It will expect a response from the
+Agent containing information about the available services.
+
+###### Agent Connection
+
+On connection, the Agent sends a CustomMessageCapabilities message including the message
+type "io.opentelemetry.find_services". In response, the Server sends a
+CustomMessageCapabilities message including the message type
+"io.opentelemetry.find_services_response".
```

Comment:
In this scenario, the server specifies `io.opentelemetry.find_services_response` which isn't one of the capabilities listed by the agent. Wouldn't we expect the agent to broadcast that its custom capabilities include both `io.opentelemetry.find_services_response` and `io.opentelemetry.find_services`?

Reply:
That's a good question and maybe a little confusing. The agent supports the request and the server supports the response. Maybe the custom capability should be separate and not based on the message, e.g. `io.opentelemetry.discovery` is supported by both client and server and the corresponding message types are implied.

Reply:
I have to admit that I also had to take a minute to figure this out at first. I like that specifying supported custom messages mirrors the way capabilities work in OpAMP and that it makes it very clear what the Agent and Server support. If we separate the capability from the message, I think it would make sense in `CustomMessageCapabilities` to indicate the capability (`io.opentelemetry.discovery`) then a list of messages the peer supports receiving as part of that capability, (`request` for the Agent and `response` for the Server) and possibly the messages it may send (`response` for the Agent and `request` for the Server).

Reply:
I like `CustomCapabilities` if we're going to say that supporting a capability implies full message support. The name lines up well with existing OpAMP terminology and is a bit more to the point.

In regards to specifying supported messages or suggesting that peers support all messages for a capability, I think that it's okay to just imply full support given that we are versioning the capabilities and considering a custom capability will be used for a very specific flow. It would also make things easier to reason about during implementation.